include/wil/com.h

```diff
@@ -3002,6 +3002,150 @@ namespace wil
 
 #endif // __IObjectWithSite_INTERFACE_DEFINED__
 
+
+// if C++17 or greater
+#if WIL_HAS_CXX_17
+#ifdef WIL_ENABLE_EXCEPTIONS
+namespace details
+{
+    template<typename>
+    struct com_enumerator_next_traits;
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*)>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*) noexcept>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+    template<typename T>
+    struct has_next
+    {
+      template <typename U = T>
+      static auto test(int) -> decltype(wistd::declval<U>()->Next(0, nullptr, nullptr), wistd::true_type{});
+
+      template <typename>
+			static auto test(...) -> wistd::false_type;
+
+			static constexpr bool value = decltype(test<T>(0))::value;
+    };
+
+    template<typename T>
+    constexpr bool has_next_v = has_next<T>::value;
+
+    template<typename Interface, wistd::enable_if_t<has_next_v<Interface*>, int> = 0>

```

Comment:
nit: I don't think the `enable_if` is doing anything here?

Reply:
i figured we'd want to sfinae this too? can remove if it's not useful.

Reply:
SFINAE on class templates _can_ be useful if you've got a number of definitions fighting each other (I've used the `void_t` trick on a number of occasions for this). In this case, it's not all that useful since the traits type shouldn't ever be used with a non-IEnum* type